### DISCUSSION TITLE
`vector-tile-builder` : Hosting of vector tile web maps in environments without Internet access

### DISCUSSION THREAD

**Discussion by yuiseki** [UNopenGIS/7#20] [Show and tell]
I would like to show you one of my repository:
- https://github.com/yuiseki/vector-tile-builder

I have today achieved to hosting of vector tile web maps in environments with no internet access.
This is one of the truly `portable` vector tile web maps now.

What I should achieved is followings:
- Using `maplibre-gl` to display vector tile web maps
- `maplibre-gl` require `style.json`, a Mapbox Style Specification format file
  - `style.json` require `tile.json`, a TileJSON format file
    - `tile.json` require zxy orderd PBF files like: `/{z}/{x}/{y}.pbf` with absolute URI
  - Also `style.json` require `sprite` and `glyphs` with absolute URI

Here is the approach I took:
- From `Geofabrik`, Download OpenStreetMap snapshot of specific region as PBF format file
  - Then I got `region.pbf`
  - Using `tilemaker`, Convert PBF format file to a MBTiles format file
    - Then I got `region.mbtiles`
    - Using `tippecanoe`, Split MBTiles format file to zxy orderd PBF files
      - Then I got `/{z}/{x}/{y}.pbf` for `tile.json`
- Using `mbtiles2tilejson`, Generate a TileJSON format file from a MBTiles format file
  - Then I got `tile.json` for `style.json`
- Using `charites`, Geenrate a Mapbox Style Specification format file
  - Then I got `style.json`
- Clone sprite and glyphs from `openmaptiles` project
  - Then I got `sprite` and `glyphs` in `style.json`

This repository contains a Dockerfile with all the required software packages.
This Dockerfile also works on Raspberry Pi 4 Model B.
Raspberry Pi 4 Model B has capability as Wi-Fi AP mode.

I am now able to connect to my Raspberry Pi, which is disconnected from the internet and acting as a Wi-Fi AP, and display the map on browser.

That is all. Please give me your feedback!

**Comment by ubukawa** [comment]
Thank you @yuiseki  for sharing your experience! 
This is something I would like to test in near future. 
Thank you very much and I will learn from your repository step by step.
(I will not install Docker in my raspberry pi 3B, but I will learn from your scripts in the repository. )


The following may not be necessary for you, but let me share my thoughts just for your information:
- (FYI 1) I do not often use tile.json, but I can directly refer to the absolute URL of PBF file in style.json by using "tiles" instead of "url" in the source. If you use tile.json only for the location of URL, please try "tiles."
   -  "url": "....../tilejson.json"
   -  "tiles": "..../{z}/{x}/{y}.pbf"  (<--This will work as well)
- (FYI 2) To save the storage and to save the data conversion process, we often use @mapbox/mbtiles to deliver pbf tile from mbtiles upon requested. (Although it is not static hosting, but we can develop it with simple nodejs script like https://github.com/un-vector-tile-toolkit/onyx/blob/master/app.js). I learned your repository https://github.com/yuiseki/mbtiles2tilejson and noticed that you do not prefer dynamic pbf hosting from mbtiles, but this is sometimes useful because we can save the storage.

Thank you once again for kindly sharing your experience. 
I will also share my experience when I have some.
